src/lib/maps/Client.ts

```diff
@@ -0,0 +1,265 @@
+
+    if (categories) {
+      const encodedCategories = encodeParameter('categories', categories.toString());
+      parameters.push(encodedCategories);
+    }
+
+    const getUrl = `${url.https}?${parameters.join('&')}`;
+    const response = await fetch(getRequest(getUrl));
+    const dataviewResponse = await response.json();
+
+    return dataviewResponse;
+  }
+
+  public async instantiateMap(mapConfig: unknown) {
+    let response;
+
+    try {
+      const payload = JSON.stringify(mapConfig);
+      response = await fetch(this.makeMapsApiRequest(payload));
+    } catch (error) {
+      throw new Error(
+        `Failed to connect to Maps API with the user ('${this._credentials.username}'): ${error}`
+      );
+    }
+
+    const layergroup = (await response.json()) as never;
+
+    if (!response.ok) {
+      this.dealWithWindshaftErrors(response, layergroup);
+    }
+
+    return layergroup;
+  }
+
+  private makeMapsApiRequest(config: string) {
+    const encodedApiKey = encodeParameter('api_key', this._credentials.apiKey);
+    const parameters = [encodedApiKey];
+    const url = this.generateMapsApiUrl(parameters);
+
+    const getUrl = `${url}&${encodeParameter('config', config)}`;
+
+    if (getUrl.length < REQUEST_GET_MAX_URL_LENGTH) {
+      return getRequest(getUrl);
+    }
+
+    return postRequest(url, config);
+  }
+
+  private dealWithWindshaftErrors(
+    response: { status: number },
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    layergroup: any
+  ) {
+    const errorForCode = errorHandlers[response.status];
+
+    if (errorForCode) {
+      errorForCode(this._credentials);
+      return;
+    }
+
+    throw new Error(`${JSON.stringify(layergroup.errors)}`);
+  }
+
+  private generateMapsApiUrl(parameters: string[] = []) {
+    const base = `${this._credentials.serverURL}api/v1/map`;
+    return `${base}?${parameters.join('&')}`;
+  }
+}
+
+export interface AggregationColumn {
+  // eslint-disable-next-line camelcase
+  aggregate_function: string;
+  // eslint-disable-next-line camelcase
+  aggregated_column: string;
+}
+
+export interface StatsColumn {
+  topCategories: number;
+  includeNulls: boolean;
+}
+
+export interface Sample {
+  // eslint-disable-next-line camelcase
+  num_rows: number;
+  // eslint-disable-next-line camelcase
+  include_columns: string[];
+}
+
+export interface MapOptions {
+  bufferSize?: BufferSizeOptions;
+  sql?: string;
+  dataset?: string;
+  vectorExtent: number;
+  vectorSimplifyExtent: number;
+  metadata?: {
+    geometryType: boolean;
+    columnStats?: StatsColumn;
+    dimensions?: boolean;
+    sample?: Sample;
+  };
+  aggregation?: {
+    placement: string;
+    resolution: number;
+    threshold?: number;
+    columns?: Record<string, AggregationColumn>;
+    dimensions?: Record<string, { column: string }>;
+  };
+}
+
+interface BufferSizeOptions {
+  png?: number;
+  'grid.json'?: number;
+  mvt?: number;
+}
+
+export interface MapInstance {
```

Comment:
👍  As this file is getting bigger, we'll make a split into several files in future PRs